gazebo/dave_ros_gz_plugins/src/DVLBridge.cc

```diff
@@ -0,0 +1,197 @@
+
+  if (!rclcpp::ok())
+  {
+    rclcpp::init(0, nullptr);
+  }
+
+  this->ros_node_ = std::make_shared<rclcpp::Node>("dvl_bridge_node");
+
+  // Grab dvl topic from SDF
+  if (!_sdf->HasElement("topic"))
+  {
+    this->dataPtr->dvl_topic = "/dvl/velocity";
+    gzmsg << "dvl topic set to default:  " << this->dataPtr->dvl_topic << std::endl;
+  }
+  else
+  {
+    this->dataPtr->dvl_topic = _sdf->Get<std::string>("topic");
+    gzmsg << "dvl topic: " << this->dataPtr->dvl_topic << std::endl;
+  }
+
+  // Gazebo subscriber
+  std::function<void(const gz::msgs::DVLVelocityTracking &)> callback =
+    std::bind(&DVLBridge::receiveGazeboCallback, this, std::placeholders::_1);
+
+  this->dataPtr->gz_node.Subscribe(this->dataPtr->dvl_topic, callback);
+
+  // ROS2 publisher
+  this->dataPtr->dvl_pub =
+    this->ros_node_->create_publisher<dave_interfaces::msg::DVL>(this->dataPtr->dvl_topic, 1);
+}
+
+void DVLBridge::receiveGazeboCallback(const gz::msgs::DVLVelocityTracking & msg)
+{
+  std::lock_guard<std::mutex> lock(this->dataPtr->mutex_);
+
+  gzmsg << "dave_ros_gz_plugins::DVLBridge::receiveGazeboCallback" << std::endl;
+
+  auto dvl_msg = dave_interfaces::msg::DVL();
+  dvl_msg.header.stamp.sec = msg.header().stamp().sec();
+  dvl_msg.header.stamp.nanosec = msg.header().stamp().nsec();
+
+  std::string dvl_type;
+  std::string target_type;
+  std::string reference_type;
+
+  switch (msg.type())
+  {
+    case DVLType::DVL_TYPE_UNSPECIFIED:
+      dvl_type = "DVL_TYPE_UNSPECIFIED";
+      break;
+    case DVLType::DVL_TYPE_PISTON:
+      dvl_type = "DVL_TYPE_PISTON";
+      break;
+    case DVLType::DVL_TYPE_PHASED_ARRAY:
+      dvl_type = "DVL_TYPE_PHASED_ARRAY";
+      break;
+    default:
+      dvl_type = "DVL_TYPE_UNSPECIFIED";
+  }
+
+  switch (msg.target().type())
+  {
+    case DVLTargetType::DVL_TARGET_UNSPECIFIED:
+      target_type = "DVL_TARGET_UNSPECIFIED";
+      break;
+    case DVLTargetType::DVL_TARGET_BOTTOM:
+      target_type = "DVL_TARGET_BOTTOM";
+      break;
+    case DVLTargetType::DVL_TARGET_WATER:
+      target_type = "DVL_TARGET_WATER";
+      break;
+    default:
+      target_type = "DVL_TARGET_UNSPECIFIED";
+  }
+
+  dvl_msg.type = dvl_type;
+
+  dvl_msg.target.type = target_type;
+  dvl_msg.target.range = msg.target().range().mean();
+
+  dvl_msg.velocity.twist.linear.x = msg.velocity().mean().x();
+  dvl_msg.velocity.twist.linear.y = msg.velocity().mean().y();
+  dvl_msg.velocity.twist.linear.z = msg.velocity().mean().z();
+
+  size_t covariance_size = msg.velocity().covariance().size();
+
+  for (size_t i = 0; i < covariance_size; i++)
+  {
+    dvl_msg.velocity.covariance[i] = msg.velocity().covariance()[i];
+  }
+
+  size_t beam_size = msg.beams().size();
+
+  for (size_t i = 0; i < beam_size; i++)
+  {
+    auto beam = dave_interfaces::msg::DVLBeam();
+    beam.id = msg.beams()[i].id();
+
+    switch (msg.beams()[i].velocity().reference())
+    {
+      case ReferenceType::DVL_REFERENCE_UNSPECIFIED:
```

Comment:
Changing `case ReferenceType::DVL_REFERENCE_UNSPECIFIED:` to `case gz::msgs::DVLKinematicEstimate::DVL_REFERENCE_UNSPECIFIED:` seems to compile alright. But not sure this is right fix.

Reply:
changed